profile/profile.js

```diff
@@ -0,0 +1,361 @@
+		
+		siteSelect.append($('<option disabled selected value="" style="display:none"> -- Select a site -- </option>'));
+		dateSelect.append($('<option disabled selected value="" style="display:none"> -- Select a date -- </option>'));
+		timeSelect.append($('<option disabled selected value="" style="display:none"> -- Select a time -- </option>'));
+
+		for (const [siteId, title] of sitesMap) {
+			siteSelect.append($('<option>', {
+				value: siteId,
+				text: title
+			}));
+		}
+
+		siteSelect.change(function() {
+			timeSelect.children('option').remove();
+			timeSelect.append($('<option disabled selected value="" style="display:none"> -- Select a time -- </option>'));
+			timeSelect.attr('disabled', true);			
+			dateSelect.children('option').remove();
+			dateSelect.append($('<option disabled selected value="" style="display:none"> -- Select a date -- </option>'));
+			dateSelect.attr('disabled', false);
+			
+			let siteId = $(this).val();
+
+			// populate dates select now based on which site was chosen
+			for (const dateString of shiftsMap.get(siteId).keys()) {
+				dateSelect.append($('<option>', {
+					value: dateString,
+					text: dateString
+				}));
+			}
+		});
+
+		dateSelect.change(function() {
+			timeSelect.children('option').remove();
+			timeSelect.append($('<option disabled selected value="" style="display:none"> -- Select a time -- </option>'));			
+			timeSelect.attr('disabled', false);
+			let siteId = siteSelect.val();
+			let dateString = $(this).val();
+
+			// populate shift times select now based on which date was chosen AND the user is not already signed up for it
+			for (const shift of shiftsMap.get(siteId).get(dateString)) {
+				if (shift.signedUp) continue;
+				timeSelect.append($('<option>', {
+					value: shift.shiftId,
+					text: `${shift.startTime} - ${shift.endTime}`
+				}));
+			}
+		});
+
+		let cancelButton = $('<button type="button"></button>').addClass("btn btn-seconday").html("Cancel").click(function(){
+			$(this).parent().remove();
+		});
+
+		let signUpButton = $('<button type="button"></button>').addClass('btn btn-primary').html('Sign Up').click(function() {
+			$(this).prop('disabled', true);
+			cancelButton.prop('disabled', true);
+
+			let siteId = siteSelect.val();
+			let dateString = dateSelect.val();
+			let shiftId = timeSelect.val();
+
+			$.ajax({
+				url: "/server/profile/profile.php",
+				type: "POST",
+				dataType: "JSON",
+				data: {
+					shiftId: shiftId,
+					action: "signUpForShift"
+				},
+				cache: false,
+				success: function(response) {
+					if (response.success) {
+						shiftRow.remove();
+						// Find the shift and append the p tags to show the user is signed up
+						for (const shift of shiftsMap.get(siteId).get(dateString)) {
+							if (shift.shiftId === shiftId) {
+								let title = sitesMap.get(siteId);
+								appendSignedUpShift(title, dateString, shift.startTime, shift.endTime, response.userShiftId);
+								break;
```

Comment:
Since you aren't calling to the database again to re-pull all of the shifts. The shift that the user just signed up for is still available to be selected. If they try to sign up for the shift again, it just gets a generic error from the server.

The shift.signedUp value should probably be updated to true here to prevent the user from being able to try signing up for the same shift twice.